### ISSUE TITLE
I am not sure if I am creating an id for each new blog or not passing it to edit.

### ISSUE THREAD

**Issue by novakw267** [ga-wdi-boston/capstone-project#585]
I managed to get the transition to the edit page working. However I don't seem to passing anything to define which blog is going to be edited. 
 
Typing in the form fields displayed gives 

```
"Assertion Failed: Cannot call set with 'title' on an undefined object."
```

This makes me think one of two things is happening.

1) when I create a blog it is just returning as an empty object
or 
2) when I pass the edit action and transition to the new view, it is not passing the Id or object to be edited. 

blog-form.hbs (this component containing form field to enter the creation of the blog)
```
<form>
<div class="form-group">
  <label>Title</label>
  {{input type="text"
          class="form-control"
          maxlength="50"
          placeholder="Title"
          value=blog.title}}
</div>

<div class="form-group">
  <label>Subject</label>
  {{input type="text"
          class="form-control"
          maxlength="75"
          placeholder="Subject"
          value=blog.subject}}
</div>

<div class="form-group">
  <label>Body</label>
  {{input type="text"
          class="form-control"
          maxlength="500"
          placeholder="Body"
          value=blog.body}}
</div>

<div class="form-group">
  <label>Author</label>
  {{input type="text"
          class="form-control"
          maxlength="50"
          placeholder="Author"
          value=blog.author}}
</div>


<button type="submit" class="btn btn-primary" {{action 'save'}}>
  Create
</button>

<button class="btn btn-default" {{action "cancel"}}>
  Cancel
</button>
</form>
```

blogs/new.hbs (the template containing the action binding for the create)
```
{{blog-form blog=model save="createBlog" cancel='cancel'}}
```

blogs/new.js (this is the routes file for creating a new blog)
```
import Ember from 'ember';

export default Ember.Route.extend({
  model () {
  return this.get('store').createRecord('blog');
},
actions: {
  createBlog(blog) {
    console.log('creating blog', blog);
    blog.save(blog)
    .then(() => this.transitionTo('blogs'));
  },
  cancel () {
    history.back();
  }
}
});
```

blog-form.js (component for saving the created blog)
```
import Ember from 'ember';

export default Ember.Component.extend({
  actions: {
    save() {
      console.log('save', this.get('blog'));
      this.sendAction('save', this.get('blog'));
      this.set('createBlog', null);
    },
    cancel() {
      this.get('blog').rollbackAttributes();
      this.sendAction('cancel');
    }
  }
});
```
 Those I believe are all the files involved in creating a new blog

I am going make a second post in this issue for the edit path to see if the ID is just not being passed and I missed something.

**Comment by raq929** [comment]
I'm not sure if we're talking about edit or create here. This seems to be a route for creating a new blog. 
Let's choose one to focus on. If we're editing, than we should be on an edit route not a `new` route.

**Comment by novakw267** [comment]
Okay so during the creation I logged what is sent when I create a blog. It logs and ember ID and an ID for the blog( I think) However in the log if I open the object and it says data is an and empty object. It sees the model, and it records to my back end server. 

I think it is a problem with edit not sending the ID again because after checking my back end server when I get the same error my back end doesn't spit anything out

**Comment by novakw267** [comment]
Templates/Components/blog-list/index.hbs(this contains the button that pulls up the edit form)
```
<h2>Title: {{blog.title}}</h2>
<h3>Subject: {{blog.subject}}</h3>
<h3>Blog: {{blog.body}}</h3>
<h3>Author: {{blog.author}}</h3>
<br>
<button class='comment btn btn-primary'>Add Comment</button>
<button class="btn btn-primary" {{action 'edit'}}>Edit</button>
<button class='delete btn btn-danger' type='submit' {{action 'delete'}}>Delete</button>
```

Templates/Components/blog-list/post.hbs (this is the edit form that appears)
```
<form>
<div class="form-group">
  <label>Title</label>
  {{input type="text"
          class="form-control"
          maxlength="50"
          placeholder="Title"
          value=blog.title}}
</div>

<div class="form-group">
  <label>Subject</label>
  {{input type="text"
          class="form-control"
          maxlength="75"
          placeholder="Subject"
          value=blog.subject}}
</div>

<div class="form-group">
  <label>Body</label>
  {{input type="text"
          class="form-control"
          maxlength="500"
          placeholder="Body"
          value=blog.body}}
</div>

<div class="form-group">
  <label>Author</label>
  {{input type="text"
          class="form-control"
          maxlength="50"
          placeholder="Author"
          value=blog.author}}
</div>


<button type="submit" class="btn btn-primary" {{action 'save'}}>
  Update
</button>

{{#link-to "blogs.index" class="btn btn-primary"}}Go Back{{/link-to}}

</form>
```

Templates/blogs/index.hbs (this is the action binding for the edit button)
```
{{#each model as |blog|}}
{{blog-list/index
  blog=blog
  edit="editBlog"
  delete="deleteBlog"
}}
{{/each}}
{{outlet}}
```

Templates/blog/edit.hbs (this has the action controller for saving the edit)

```
{{blog-list/post
  blog=blog
  save=createBlog
  cancel=cancel
}}
{{outlet}}
```

routes/blogs/index.js (this is the routes for the editBlog)

```
import Ember from 'ember';

export default Ember.Route.extend({
  // model () {
  //   console.log('k');
  //   return this.get('store').findAll('blog');
  // },
  actions: {
  editBlog (blog) {
    console.log('edit2', blog);
    this.transitionTo('blog.edit', blog);
  },
  deleteBlog (blog) {
    console.log("gone!");
    blog.destroyRecord();
  },
},
});
```

Routes/blog/edit.js (this is the route for the saveBlog)

```
import Ember from 'ember';

export default Ember.Route.extend({
  actions: {
    editBlog (blog) {
      this.transitionTo('blog-list.post', blog);
    },
  saveBlog(blog) {
    blog.save()
    .then(() => this.transitionTo('blogs'));
  },
  deleteBlog (blog) {
    console.log("gone!");
    blog.destroyRecord();
  },
  cancel () {
    console.log('go back');
    history.back();
  }
}
});
```

Components/blog-list/index.js (component for edit function)

```
import Ember from 'ember';

export default Ember.Component.extend({
  actions: {
    edit() {
      console.log('edit 3');
      this.sendAction('edit', this.get('blog'));
    },
    delete() {
      console.log('almost');
      this.sendAction('delete', this.get('blog'));
    },
  }
});
```

Components/blog-list/post.js (component for the save function)

```
import Ember from 'ember';

export default Ember.Component.extend({
  actions: {
    edit() {
      console.log('edit 3');
      this.sendAction('edit', this.get('blog'));
    },
    delete() {
      console.log('almost');
      this.sendAction('delete', this.get('blog'));
    },
  }
});
```

**Comment by novakw267** [comment]
So in order to solve this we had to change the router I had the ID being called twice

```
this.route('blog', { path: '/blogs/:blog_id' }, function() {
      this.route('edit', { path: '/edit'});
  });
``` 
We removed the extra id call in the edit path.

After which we also had to change the binding in the template file to 

```
blog=model
```
instead of 
```
blog=blog
```
 
That fixed being able to pass the ID to the edit form. Which in turn allowed for the changes to be assigned to that post and saved.
